src/storage-client/src/storage_collections.rs

```diff
@@ -0,0 +1,2640 @@
+        identifiers: Vec<GlobalId>,
+    );
+
+    /// Assigns a read policy to specific identifiers.
+    ///
+    /// The policies are assigned in the order presented, and repeated
+    /// identifiers should conclude with the last policy. Changing a policy will
+    /// immediately downgrade the read capability if appropriate, but it will
+    /// not "recover" the read capability if the prior capability is already
+    /// ahead of it.
+    ///
+    /// This [StorageCollections] may include its own overrides on these
+    /// policies.
+    ///
+    /// Identifiers not present in `policies` retain their existing read
+    /// policies.
+    fn set_read_policies(&mut self, policies: Vec<(GlobalId, ReadPolicy<Self::Timestamp>)>);
+
+    /// Acquires and returns the earliest possible read holds for the specified
+    /// collections.
+    fn acquire_read_holds(
+        &mut self,
+        desired_holds: Vec<GlobalId>,
+    ) -> Result<Vec<ReadHold<Self::Timestamp>>, ReadHoldError>;
+
+    /// Applies `updates` and sends any appropriate compaction command.
+    ///
+    /// This is a legacy interface that should _not_ be used! It is only used by
+    /// the compute controller.
+    fn update_read_capabilities(
+        &mut self,
+        updates: &mut BTreeMap<GlobalId, ChangeBatch<Self::Timestamp>>,
+    );
+}
+
+#[derive(Debug, Clone)]
+enum PersistTxns<T> {
+    EnabledEager,
+    EnabledLazy { txns_read: TxnsRead<T> },
+}
+
+impl<T: TimelyTimestamp + Lattice + Codec64> PersistTxns<T> {
+    fn expect_enabled_lazy(&self, txns_id: &ShardId) -> &TxnsRead<T> {
+        match self {
+            PersistTxns::EnabledLazy { txns_read, .. } => {
+                assert_eq!(txns_id, txns_read.txns_id());
+                txns_read
+            }
+            PersistTxns::EnabledEager { .. } => {
+                panic!("set if txns are enabled and lazy")
+            }
+        }
+    }
+}
+
+/// Implementation of [StorageCollections] that is shallow-cloneable and uses a
+/// background task for doing work concurrently, in the background.
+#[derive(Debug, Clone)]
+pub struct StorageCollectionsImpl<
+    T: TimelyTimestamp + Lattice + Codec64 + From<EpochMillis> + TimestampManipulation,
+> {
+    /// The fencing token for this instance of [StorageCollections], and really
+    /// all of the controllers and Coordinator.
+    envd_epoch: NonZeroI64,
+
+    /// The set of [ShardIds](ShardId) to finalize.
+    ///
+    /// This is a separate set from `finalized_shards` because we know that some
+    /// environments have many, many finalizable shards that we are struggling
+    /// to finalize.
+    finalizable_shards: Arc<std::sync::Mutex<BTreeSet<ShardId>>>,
+
+    /// The set of [ShardIds](ShardId) we have finalized.
+    ///
+    /// This is a separate set from `finalizable_shards` because we know that
+    /// some environments have many, many finalizable shards that we are
+    /// struggling to finalize.
```

Comment:
I'm not sure this explains the fact that we have both very well...`finalized_shards` is in fact cleared periodically, right?

Reply:
I copied these as-is from `StorageController`, but yes, the explanation is not super clear. I'll give it a think.

Reply:
thank you!